File: 147.insertion-sort-list.java
/*
 * @lc app=leetcode id=147 lang=java
 *
 * [147] Insertion Sort List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode();
        ListNode prev = dummy;

        while (head != null) {
            ListNode temp = head.next;
            
            // reset the last node of the result list to the start
            // when the current node is lower than the last (largest) node
            if (prev.val > head.val) {
                prev = dummy;
            }

            while (prev.next != null && prev.next.val < head.val) {
                prev = prev.next;
            }

            head.next = prev.next;
            prev.next = head;
            head = temp;
        }

        return dummy.next;
    }
}
// @lc code=end

File: 147.insertion-sort-list.py
#
# @lc app=leetcode id=147 lang=python
#
# [147] Insertion Sort List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        dummy = ListNode()
        prev = dummy

        while head:
            temp = head.next

            if (prev.val > head.val):
                prev = dummy
            
            while prev.next and prev.next.val < head.val:
                prev = prev.next
            
            head.next = prev.next
            prev.next = head
            head = temp
        
        return dummy.next
# @lc code=end

